[KL-502] Cache requests + Rate limitting

LikoIlya opened the pull request:
Add cache for routes
Add rate limits 20 req/3s

LikoIlya committed:
[KL-501] implement building backend image

LikoIlya committed:
Upd description

to trigger build image

LikoIlya committed:
Add quotes for workflow

LikoIlya committed:
Update workflow

LikoIlya committed:
Update labeling

LikoIlya committed:
fix naming

LikoIlya committed:
Update branch

LikoIlya committed:
Add cache + separating routes

LikoIlya committed:
Apply rate limits after cache

LikoIlya committed:
Move back rate limitter MW + linting

LikoIlya committed:
semi

LikoIlya committed:
Refactoring

LikoIlya committed:
Refactor rate limit part

unixvb commented on apps/backend/src/apps/debank/utils/rate-limit.ts:
```diff
@@ -0,0 +1,28 @@
+const formStat = (key: string, lrRes: RateLimiterRes): string =>
+  `${key} c.:${lrRes.consumedPoints} pts, l.: ${lrRes.remainingPoints} pts. Reset in ${lrRes.msBeforeNext / 1000}s`;
+
+export const rateLimiterMiddleware = (req: Request, res: Response, next: NextFunction) => {
```
I know that we don't have anything else right now, but could we set this rate limit "globaly", not only for debank api?  
For example, you could add a health-check endpoint and make a single rate limit for debank and this endpoint.

LikoIlya replied:
we can apply this middleware for any route we want e.g. /*, /debank/*, etc, or use another instance of `RateLimiterRedis` to apply the other rate-limit policy and handling. I can try to implement this middleware as global util and use it in routes. But what config should we pass to the middleware?

unixvb replied:
the same config as we have for `debank` right now

unixvb commented on apps/backend/src/config/redis/guards.ts:
```diff
@@ -0,0 +1,28 @@
+import { RedisConfig } from './types';
+
+const redisHostGuard = (env = process.env): string => {
+  if (!('REDIS_HOST' in env)) {
+    throw new Error('REDIS_HOST not defined');
+  }
+
+  return String(env.REDIS_HOST);
+};
+const redisPortGuard = (env = process.env): number => Number(env.REDIS_PORT ?? 6379);
+
+export const redisConfigGuard = (env = process.env): RedisConfig => {
+  const config: RedisConfig = {
+    host: redisHostGuard(env),
+    port: redisPortGuard(env),
+    enable_offline_queue: false
+  };
+  if ('REDIS_PASSWORD' in env) {
+    config.password = env.REDIS_PASSWORD;
```
add new empty env variables into `apps/backend/.env.dist` file

unixvb commented on apps/backend/src/apps/debank/v1/user/history-list.ts:
```diff
@@ -0,0 +1,31 @@
+};
+
+const getUserHistoryList = (req: Request, res: Response, next: NextFunction) =>
+  proxyDeBankRequest<interfaces.ActivityResponse>('user/history_list', req.query)
```
could you change shared interfaces imports to this:
`import { ActivityResponse } from 'backend-types';`  
and use them without `interfaces.`prefix `proxyDeBankRequest<ActivityResponse>(`

unixvb commented on libs/backend-types/src/interfaces/activity-response.interface.ts:
```diff
@@ -0,0 +1,33 @@
+  site_url: string;
+}
+
+export interface ActivityResponse {
```
reuse this types in frontend

LikoIlya committed:
Fix get redis conf from singleton

add values to env.dist

LikoIlya committed:
add type validation

Usage of backend types on ui,
move rate-limit and cache to global utils

LikoIlya committed:
Merge branch 'development' into KL-502/cache-requests+rate-limit

LikoIlya committed:
Add validators to debank requests api

unixvb commented on libs/ui/src/api/debank.ts:
```diff
@@ -62,7 +66,12 @@ export const getTokenList = memoize(
+        params: {
+          id: publicKeyHash,
+          chain_id: chainId
+        } as ChainUserRequest
```
why do you need all these assertions in this file?  
try to remove them

LikoIlya committed:
Remove type assertions

LikoIlya committed:
Resolve conflicts with develop branch

# Conflicts:
#	libs/ui/src/api/debank/debank.ts
#	libs/ui/src/api/debank/types.ts
#	libs/ui/src/store/wallet/wallet.actions.ts

LikoIlya committed:
Merge branch 'development' into KL-502/cache-requests+rate-limit

LikoIlya committed:
remove unused after merge

lourenc commented on apps/backend/src/apps/debank/v1/user/history-list.ts:
```diff
@@ -0,0 +1,43 @@
+  validateRequestMiddleware
+];
+
+const expire = {
```
I guess it would be better to have the type information here to be consistent across the codebase:
`const expire: ExpirationConfig`

lourenc reviewed (changes requested):
The only minor issue

LikoIlya committed:
Add typing for ExpirationConfig in history-list.ts

lourenc reviewed (approved):
👍
